### PR TITLE
Fix position embeddings for GPT-J and CodeGen

### PR DESCRIPTION
**Pull request by njhill**
# What does this PR do?

Identical inputs to GPT-J and CodeGen models will currently generate different outputs if they are padded differently (for example in a batch of variable sequence lengths).

This PR reverts the recent change #21869 that removes GPT-J `position_ids`, and then applies similar changes as were done for GPT-J XLA in #17986.

~One copy of the precomputed position embeddings is shared between all of the layers.~

Related issue: #21080

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@gante

### CODE REVIEW

**Comment by njhill** [comment]
I have tested this with my own code/usecase but wanted to check that there is interest in the contribution before also updating any applicable unit tests.

I also wonder whether there should be a universal test applied to all models that just tests the same input with different amounts of padding and makes sure that the output is identical?

File: src/transformers/models/gptj/modeling_gptj.py
Comment (gante) [reply src/transformers/models/gptj/modeling_gptj.py:64]: (would be copying from itself)

**Review by gante — COMMENTED** [review]
A few comments on the GPT-J side. Once we're both happy with GPT-J, Codegen should be copy/paste :)

File: src/transformers/models/gptj/modeling_gptj.py
Comment (njhill) [reply src/transformers/models/gptj/modeling_gptj.py:64]: Oops, I actually had made the changes to codegen first and copied in the other direction :)

File: src/transformers/models/gptj/modeling_gptj.py
Comment (njhill) [reply src/transformers/models/gptj/modeling_gptj.py:102]: I've memoized the precomputed positions at the model level to avoid generating/storing a redundant copy for every layer. Perhaps it's small enough (a few MB I think) that it's not a concern? If so I can revert this.

I think we also want these to be on the target device which I wasn't sure is necessarily known at construction time, hence the lookup taking a device from generation-time tensors.

Comment (gante) [reply src/transformers/models/gptj/modeling_gptj.py:102]: Agreed that is more efficient! However, in `transformers`, we prioritize consistent APIs and classes over efficiency, and GPT-J's classes would become an outlier with that extra init parameter :D

File: src/transformers/models/gptj/modeling_gptj.py
Comment (njhill) [inline src/transformers/models/gptj/modeling_gptj.py:197]: @gante let me know if there's a better way to handle this

Comment (gante) [reply src/transformers/models/gptj/modeling_gptj.py:207]: LGTM 👍

File: src/transformers/models/gptj/modeling_gptj.py
Comment (gante) [inline src/transformers/models/gptj/modeling_gptj.py:209]: Uhmmm can we cast to long in `GPTJModel`? (so as to do it only once per forward pass)

**Review by gante — APPROVED** [review]
Looks great!

Two additional requests before merging:
1 - Can you confirm that existing slow tests pass for GPT-J and Codegen? They are tests with `batch_size=1`, so they should see no changes.
2 - Can you add a slow integration test on Codegen with `batch_size=2`? (GPT-J only has a massive model, so our CI doesn't run its integration tests :( ). Preferably, with an example that changes with the introduction of these changes, if you can find one 💛

**Comment by ydshieh** [comment]
@gante I didn't review this PR, but I see this is related to issue #21080, and therefore to PR #21853 indirectly, which was reverted in #22093 due to some unexpected tests failure (PT/TF, PT/Flax).

So before merging this PR, it's better to verify the cross tests, as well as the slow tests too (always better).

**Comment by gante** [comment]
Thanks for the heads up @ydshieh! 🙏 

I'll make sure all related slow tests (and the tests that failed after merging #21853 ) are passing before merging.

**Comment by njhill** [comment]
Thanks @gante ... I'm kind of new to this but will figure out how to verify/update the tests per your request.

The main problem I've run into though is newly-failing `torch.fx` tracing [tests](https://app.circleci.com/pipelines/github/huggingface/transformers/59686/workflows/0fe3ef82-316a-482d-802d-d245028b2bf6/jobs/730191/parallel-runs/0/steps/0-111):
```
FAILED tests/models/gptj/test_modeling_gptj.py::GPTJModelTest::test_torch_fx - AssertionError: Couldn't trace module: Proxy object cannot be iterated. This can be attempted when the Proxy is used in a loop or as a *args or **kwargs function argument. See the torch.fx docs on pytorch.org for a more detailed explanation of what types of control flow can be traced, and check out the Proxy docstring for help troubleshooting Proxy iteration errors
FAILED tests/models/gptj/test_modeling_gptj.py::GPTJModelTest::test_torch_fx_output_loss - AssertionError: Couldn't trace module: Proxy object cannot be iterated. This can be attempted when the Proxy is used in a loop or as a *args or **kwargs function argument. See the torch.fx docs on pytorch.org for a more detailed explanation of what types of control flow can be traced, and check out the Proxy docstring for help troubleshooting Proxy iteration errors
```

I've tried some different variations to the logic but always end up with similar kind of errors. I think it may stem from the [index_select](https://github.com/huggingface/transformers/pull/22069/files#diff-61155574bf9c9669ccdfdf7dd508a5979b4e4915cc95f7ff4a63fee05a0e2715R209) operation. Any pointers/ideas would be appreciated!

**Review by sgugger — APPROVED** [review]
Thanks for working on this. Before merging, let's make sure the cross tests pass though this time ;-)

**Comment by gante** [comment]
Hey @njhill 👋 

I've tried to fix the issue you mentioned with no success. It seems like we are between a rock and a hard place -- the changes you made, by design, make `sincos` dependent on the values of `position_ids`. In other hand, `sincos` becomes a tensor impossible to predict at compile time with `torch.fx`, i.e. dynamic tensor. Ultimately, no matter how we rewrite the code (AFAIK), we will hit this barrier, causing the test to fail.

@sgugger @fxmarty is there a way we can make `torch.fx` ignore a function? (or do you have suggestions?) The change in this PR makes GPT-J correct in the presence of left-padding, but breaks compatibility with `torch.fx` 🙈 

(Pastebin containing the code with modifications, working through the exceptions until I got stuck: https://pastebin.com/T0HpD07C)

**Comment by sgugger** [comment]
Also cc @michaelbenayoun for torch fx.

**Comment by njhill** [comment]
Thanks @gante, it sounds like you followed a similar path to me w.r.t. trying different arrangements of the logic to get around this. I was guessing this couldn't be the only occurrence of this dynamic tensor issue in the library - is dynamic slicing done elsewhere and if so how does it work with `torch.fx`?

**Comment by michaelbenayoun** [comment]
Hi @njhill,

The issue here (from what I could understand from [this](https://app.circleci.com/pipelines/github/huggingface/transformers/59686/workflows/0fe3ef82-316a-482d-802d-d245028b2bf6/jobs/730191/parallel-runs/0/steps/0-111)), seems to be that during tracing we do not have regular tensors but rather symbolic "proxies".

In the following code we are trying to call `__iter__` on `sincos` which is symbolic, we do not know its length (again, not 100% sure but guessing).

```python
sincos = [t.contiguous() for t in sincos]
```

But the previous line is :
```python
sincos = torch.split(sincos, sincos.shape[-1] // 2, dim=-1)
```
Meaning that the list has:

- 2 elements if `sincos.shape[-1]` is an even number
- 3 elements if `sincos.shape[-1]` is an odd number.

So could you try this:

```python
sincos = torch.split(sincos, sincos.shape[-1] // 2, dim=-1)
len_sincos = 2 + torch.remainder(torch.tensor(sincos.shape[-1], 2))
sincos = [sincos[idx].contiguous() for idx in torch.arange(len_sincos)]
```

Tell me if this works!

**Comment by njhill** [comment]
Thanks @michaelbenayoun. You are right that this seems to be the fact that a symbolic proxy tensor is introduced somewhere, however I think that this stems from the tensor-based indexing here:
```python
sincos = embed_positions[position_ids]
```
The proxy iterator errors are easy to circumvent but just move the problem until later where (inevitably?) the size of the proxy tensor is used for flow control. I've pushed a couple of small updates to the PR to demonstrate this... you can see the latest error in the tests [here](https://app.circleci.com/pipelines/github/huggingface/transformers/59903/workflows/203a46e5-1bc2-4ab0-9085-4992384db930/jobs/733587). As @gante pointed out above:
> Ultimately, no matter how we rewrite the code (AFAIK), we will hit this barrier, causing the test to fail.

Could we at least make this path conditional such that it isn't followed in the `torch.fx` case, i.e. declare that variable padding is unsupported in that case?

**Comment by gante** [comment]
Hey @njhill -- I think the conditional path is a sensible idea, at least for now (we can always revisit it later). #22161 reports a similar problem on another demanded model, so I would like to merge the fix as soon as possible 🤗 

For context, other places in the `transformers` do it. Check [here](https://github.com/huggingface/transformers/blob/f7329751fe5c43365751951502c00df5a4654359/src/transformers/models/t5/modeling_t5.py#L845) for an example.

**Comment by michaelbenayoun** [comment]
@njhill The HF tracer is supposed to keep track of "concrete" metadata during tracing to allow for that. 
In this case, either this does not work with `len`, which is possible (I do not remember tbh), or it means than an op does not support the meta device, hence breaking the concrete metadata accumulation.

Since in this case you are trying to check the rank of the tensor, could you try replacing `len(tensor.shape)` by `tensor.ndim`?

**Comment by njhill** [comment]
Thanks @michaelbenayoun .. the `len` problem can be avoided by adding `torch.fx.wrap('len')`, which I'd done in the prior commit but removed in this latest commit since it seemed futile (just moving the error slightly later). So I was instead attempting to bypass the position_ids fix in the `torch.fx` case per [this comment](https://github.com/huggingface/transformers/pull/22069#issuecomment-1469690762) (so far unsuccessfully).

The problem encountered after working around the `len` problem can be seen [here](https://app.circleci.com/pipelines/github/huggingface/transformers/59903/workflows/203a46e5-1bc2-4ab0-9085-4992384db930/jobs/733587):
```
>       if len(tensor.shape) == 5:

AssertionError: Couldn't trace module: symbolically traced variables cannot be used as inputs to control flow
```
basically this traced length value is then used in a control flow condition.

**Comment by njhill** [comment]
@gante @michaelbenayoun I've got torch.fx to work with the changes now by using `torch.gather` instead of tensor based indexing and adding a couple of new tensor methods to the metadata tracking in `fx.py`.

Also rebased on latest main branch since some other CI tests started to fail I think related to a recently-merged unrelated change.

I will look into the requested additional tests next when I get a chance.

**Review by michaelbenayoun — APPROVED** [review]
LGTM

File: src/transformers/models/gptj/modeling_gptj.py
Comment (gante) [inline src/transformers/models/gptj/modeling_gptj.py:216]: Couldn't the first branch (`embed_positions = get_embed_positions(self.embed_positions, position_ids)`) work in both cases?

It would result in simpler code :)

(ditto for codegen)

**Review by gante — APPROVED** [review]
Great work sorting the `torch.fx` issues! 💛 

I left a final question/suggestion, but happy to merge it as is. @njhill I leave the final decision up to you

**Comment by gante** [comment]
For our future reference, here's a snippet that shows that left-padding is fixed with these changes:

```py
from transformers import AutoModelForCausalLM, AutoTokenizer
import torch

tok = AutoTokenizer.from_pretrained("EleutherAI/gpt-j-6B", padding_side="left")
model = AutoModelForCausalLM.from_pretrained("EleutherAI/gpt-j-6B", torch_dtype=torch.bfloat16).to(0)
tok.pad_token = tok.eos_token
model.generation_config.pad_token_id = model.generation_config.eos_token_id

inputs_1 = tok(["The brown fox"], return_tensors="pt", padding=True).to(0)
out_1 = model(**inputs_1)
out_2 = model(**inputs_1)

position_ids = torch.cumsum(inputs_1.attention_mask, dim=-1) - 1
out_3 = model(**inputs_1, position_ids=position_ids + 8)

inputs_2 = tok(["The brown fox"], return_tensors="pt", padding="max_length", max_length=10).to(0)
out_4 = model(**inputs_2)

position_ids = torch.cumsum(inputs_2.attention_mask, dim=-1) - 1
position_ids.masked_fill_(inputs_2.attention_mask == 0, 1)
out_5 = model(**inputs_2, position_ids=position_ids)

# calls with the same inputs get the same logits
print(torch.max(torch.abs(out_1.logits[:, -1, :] - out_2.logits[:, -1, :]))) # tensor(0., device='cuda:0', grad_fn=<MaxBackward1>)

# changing the position_ids changes the logits
print(torch.max(torch.abs(out_1.logits[:, -1, :] - out_3.logits[:, -1, :]))) # tensor(0.0625, device='cuda:0', grad_fn=<MaxBackward1>)

# padding and not passing position ids -> incorrect position ids -> output differences
print(torch.max(torch.abs(out_1.logits[:, -1, :] - out_4.logits[:, -1, :]))) # tensor(0.0625, device='cuda:0', grad_fn=<MaxBackward1>)

# left-padding has a much smaller impact (NOTE: setting e.g. `max_length=20` will cause the next diff to be non-zero.
# Numerical masking is not perfect :) )
print(torch.max(torch.abs(out_1.logits[:, -1, :] - out_5.logits[:, -1, :]))) # tensor(0., device='cuda:0', grad_fn=<MaxBackward1>)
```

File: src/transformers/models/gptj/modeling_gptj.py
Comment (njhill) [reply src/transformers/models/gptj/modeling_gptj.py:216]: Thanks @gante, my concern is that the cached position embeddings will initially be in CPU mem and so will need to be copied to GPU every time. The logic in `_get_embed_positions` copies the first time and then no longer needs to, but I couldn't get it to work for the `torch.fx` case, hence this if/else.

I'll add a comment to the code to make that clear. Codegen is slightly simpler because it doesn't support `torch.fx` anyhow.

**Comment by gante** [comment]
@njhill fantastic work with the `torch.fx`, I really appreciated your effort 🤗

**Comment by stas00** [comment]
This PR isn't backward compatible. It breaks with pytorch-1.8:

```
E             File "/mnt/nvme0/code/huggingface/transformers-master/src/transformers/models/gptj/modeling_gptj.py", line 63, in <module>
E               @torch.fx.wrap
E           AttributeError: module 'torch' has no attribute 'fx'
```

**Comment by ydshieh** [comment]
> This PR isn't backward compatible. It breaks with pytorch-1.8:
> 
> ```
> E             File "/mnt/nvme0/code/huggingface/transformers-master/src/transformers/models/gptj/modeling_gptj.py", line 63, in <module>
> E               @torch.fx.wrap
> E           AttributeError: module 'torch' has no attribute 'fx'
> ```
> 
> not sure if you want to revert this or have an idea how to overcome this quickly.

@stas00 

FYI, see #22291, although that PR and this PR is not directly related from the beginning when they are opened.

**Comment by stas00** [comment]
ok, the deepspeed CI is running pt-1.8 - how do we solve that then?

**Comment by ydshieh** [comment]
> ok, the deepspeed CI is running pt-1.8 - how do we solve that then?

I just saw

https://github.com/microsoft/DeepSpeed/pull/3082

opened 2 hours ago. I am not sure what will go, but I will try to follow tomorrow morning.

**Comment by stas00** [comment]
oh, ok, I guess everything is fine then. thank you for the heads up, @ydshieh

**Comment by stas00** [comment]
it still fails with pt-1.9

1. you need `import torch.fx`

2. it then fails with:

```
E             File "/mnt/nvme0/code/huggingface/transformers-master/src/transformers/models/gptj/modeling_gptj.py", line 61, in create_sinusoidal_positions
E               return torch.concat((torch.sin(sinusoid_inp), torch.cos(sinusoid_inp)), dim=1)
E           AttributeError: module 'torch' has no attribute 'concat'
```

**Comment by njhill** [comment]
Oops, I guess we should use `torch.cat()` instead

**Comment by stas00** [comment]
and it fails w/o `import torch.fx`

```
E             File "/mnt/nvme0/code/huggingface/transformers-master/examples/pytorch/language-modeling/run_clm.py", line 412, in main
E               model = AutoModelForCausalLM.from_pretrained(
E             File "/mnt/nvme0/code/huggingface/transformers-master/src/transformers/models/auto/auto_factory.py", line 470, in from_pretrained
E               model_class = _get_model_class(config, cls._model_mapping)
E             File "/mnt/nvme0/code/huggingface/transformers-master/src/transformers/models/auto/auto_factory.py", line 360, in _get_model_class
E               supported_models = model_mapping[type(config)]
E             File "/mnt/nvme0/code/huggingface/transformers-master/src/transformers/models/auto/auto_factory.py", line 602, in __getitem__
E               return self._load_attr_from_module(model_type, model_name)
E             File "/mnt/nvme0/code/huggingface/transformers-master/src/transformers/models/auto/auto_factory.py", line 616, in _load_attr_from_module
E               return getattribute_from_module(self._modules[module_name], attr)
E             File "/mnt/nvme0/code/huggingface/transformers-master/src/transformers/models/auto/auto_factory.py", line 561, in getattribute_from_module
E               if hasattr(module, attr):
E             File "/mnt/nvme0/code/huggingface/transformers-master/src/transformers/utils/import_utils.py", line 1109, in __getattr__
E               module = self._get_module(self._class_to_module[name])
E             File "/mnt/nvme0/code/huggingface/transformers-master/src/transformers/utils/import_utils.py", line 1121, in _get_module
E               raise RuntimeError(
E           RuntimeError: Failed to import transformers.models.gptj.modeling_gptj because of the following error (look up to see its traceback):
E           module 'torch' has no attribute 'fx'

```

**Comment by stas00** [comment]
I confirm that it works with `torch.cat`

perhaps use `torch.concat` but add an alias:

```
if not getattr(torch, "concat"):
    torch.concat = torch.cat
```

**Comment by stas00** [comment]
`import torch.fx` is a must - even with pt-1.10 it won't work w/o it.

**Comment by stas00** [comment]
@njhill, are you on top of fixing this?

This is a bit urgent since Deepspeed CI uses our code to test deepspeed. and currently their CI breaks.

**Comment by njhill** [comment]
@stas00 apologies I am AFK right now but could do it in a few hours. Feel free to do in the meantime if you like!

I don’t see any downside to just using `torch.cat` since it’s already an alias.

Where is it that we need to add the extra `torch.fx` import?

**Comment by stas00** [comment]
sure, I will fire off a PR - thank you for letting me know your preferences, @njhill

**Comment by stas00** [comment]
the fix is here https://github.com/huggingface/transformers/pull/22325

**Comment by stas00** [comment]
the fix has been merged.

**Pull request merged**
